components/toast_wrapper/toast_wrapper.tsx

```diff
@@ -2,103 +2,120 @@
+    scrollToNewMessage: () => void;
+    scrollToLatestMessages: () => void;
+    scrollToUnreadMessages?: () => void;
+    updateLastViewedBottomAt: (lastViewedBottom?: number) => void;
+    showSearchHintToast?: boolean;
+    onSearchHintDismiss?: () => void;
+    shouldStartFromBottomWhenUnread?: boolean;
+    isNewMessageLineReached?: boolean;
+    unreadScrollPosition?: string;
+    intl: IntlShape;
+    match: {
+        params: {
+            team?: string;
+        };
     };
-
+    actions: {
+        updateToastStatus: (toastPresent: boolean) => void;
+    };
+};
+
+type State = {
+    unreadCount?: number;
+    unreadCountInChannel: number;
+    channelMarkedAsUnread?: boolean;
+    lastViewedAt?: number;
+    showUnreadToast?: boolean;
+    showNewMessagesToast?: boolean;
+    showMessageHistoryToast?: boolean;
+    showUnreadWithBottomStartToast?: boolean;
+};
+
+export type Props = OwnProps;
+
+export class ToastWrapper extends React.PureComponent<Props, State> {
+    mounted: boolean;
     static defaultProps = {
         focusedPostId: '',
     };
 
-    constructor(props) {
+    constructor(props: Props) {
         super(props);
+        this.mounted = true;
         this.state = {
             unreadCountInChannel: props.unreadCountInChannel,
         };
     }
 
-    static countNewMessages = (postListIds, rootPosts, isCollapsedThreadsEnabled) => {
+    static countNewMessages = (postListIds: string[], rootPosts: IDMappedObjects<Post>, isCollapsedThreadsEnabled: boolean) => {
         const mark = getNewMessageIndex(postListIds);
         if (mark <= 0) {
             return 0;
         }
-        let newMessages = postListIds.slice(0, mark).filter((id) => !isIdNotPost(id));
+        let newMessages = postListIds.slice(0, mark).filter((id: string) => !isIdNotPost(id));
         if (isCollapsedThreadsEnabled) { // in collapsed mode we only count root posts
-            newMessages = newMessages.filter((id) => rootPosts[id]);
+            newMessages = newMessages.filter((id: string) => rootPosts[id]);
         }
         return newMessages.length;
     }
 
-    static getDerivedStateFromProps(props, prevState) {
+    static getDerivedStateFromProps(props: Props, prevState: State) {
         let {showUnreadToast, showNewMessagesToast, showMessageHistoryToast, showUnreadWithBottomStartToast} = prevState;
         let unreadCount;
 
         if (props.atLatestPost) {
             if (props.unreadScrollPosition === Preferences.UNREAD_SCROLL_POSITION_START_FROM_NEWEST && prevState.unreadCountInChannel) {
                 unreadCount = prevState.unreadCountInChannel + props.newRecentMessagesCount;
             } else {
-                unreadCount = ToastWrapper.countNewMessages(props.postListIds, props.rootPosts, props.isCollapsedThreadsEnabled);
+                unreadCount = ToastWrapper.countNewMessages(props.postListIds, props.rootPosts as IDMappedObjects<Post>, props.isCollapsedThreadsEnabled as boolean);
```

Comment:
As per my previous comment, if you change `rootPosts` to `Record<string, boolean>` and make it required (i.e. remove `?`) in Props type, you don't need to cast as `IDMappedObjects<Post>`

For `props.isCollapsedThreadsEnabled`, you just need to make it optional with `?`. Please see my suggestion above for `countNewMessages` definition

```suggestion
                unreadCount = ToastWrapper.countNewMessages(props.postListIds, props.rootPosts, props.isCollapsedThreadsEnabled);
```